# Import Library

In [27]:
import pandas as pd

# Data Understanding
Dataset movie_metadata adalah dataset IMDB movie yang berisi 5000 baris dengan 28 kolom, dataset di unduh dari website Kaggle. Kita akan mencoba membuat sistem rekomendasi film kepada Users sesuai genre, jika sebelumya Users menonton film dengan genre Horror maka sistem akan merekomendasikan kepada Users pilihan berbagai film bergenre Horror

In [28]:
movie_data = pd.read_csv('movie_metadata.csv')
movie_data

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
5038  ...                  6.0  English   Canada             NaN          NaN   
5039  ...                359.0  English      USA           TV-14          NaN   
5040  ...                  3.0  English      USA             NaN       1400.0   
5041  ...                  9.0  English      USA           PG-13          NaN   
5042  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

Kita lihat di atas jumlah total data kita adalah 5043 baris dengan 28 kolom, kemudian kita akan mencari tau info lebih detail data dengan fungsi info().

# Univariate Exploratory

In [29]:
movie_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

Setelah kita jalankan dengan fungsi info(), kita bisa melihat semua variable pada dataser. Di dalamnya terdiri 13 tipe data **Float**, 3 tipe data **integer**, dan 12 tipe data **Object**.

In [30]:
movie_data.isnull().sum()

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               303
budget                       492
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

Kemudian kita mencari tau apakah seluruh data memiliki Missing Value di dalamnya, setelah di cek dengan fungsi isnull().sum() lumayan banyak variable yang memiliki Missing Value.


In [31]:
movie_data.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

Kita juga akan melihat lebih detail lagi dataset kita menggunakan fungsi describe(). terlihat bermacam-macam nilai pada variable dataset kita, namun karena kita akan membuat sistem rekomendasi film berdasarkan genre, kita tidak perlu banyak variable. Kita akan mengambil variable yang di butuhkan saja yaitu, genres, movie_title, num_user_for_reviews.


In [32]:
drop = ['color', 'director_name', 'num_critic_for_reviews', 'duration', 'director_facebook_likes',
                  'actor_3_facebook_likes', 'actor_2_name', 'actor_1_facebook_likes', 'gross', 'actor_1_name',
                  'num_voted_users', 'cast_total_facebook_likes', 'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
                  'movie_imdb_link', 'language', 'country', 'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
                  'imdb_score', 'aspect_ratio', 'movie_facebook_likes']

movie_data.drop(drop, axis=1, inplace=True)


Kita membuat variable drop untuk membuang banyak atribut di dalam dataset kita, dengan begitu dataset kita menjadi sangat simple tapi tetap berfungsi dengan baik.

Setelah variable drop dengan isi atribute yang ingin kita buang telah di buat, maka kita juga akan menghapus variable tersebut dari dataset menggunakan fungsi drop().

In [33]:
movie_data

genres  \
0     Action|Adventure|Fantasy|Sci-Fi   
1            Action|Adventure|Fantasy   
2           Action|Adventure|Thriller   
3                     Action|Thriller   
4                         Documentary   
...                               ...   
5038                     Comedy|Drama   
5039     Crime|Drama|Mystery|Thriller   
5040            Drama|Horror|Thriller   
5041             Comedy|Drama|Romance   
5042                      Documentary   

                                            movie_title  num_user_for_reviews  
0                                               Avatar                 3054.0  
1             Pirates of the Caribbean: At World's End                 1238.0  
2                                              Spectre                  994.0  
3                                The Dark Knight Rises                 2701.0  
4     Star Wars: Episode VII - The Force Awakens    ...                   NaN  
...                                                 ...                   ...  
5038                           Signed Sealed Delivered                    6.0  
5039                         The Following                              359.0  
5040                              A Plague So Pleasant                    3.0  
5041                                  Shanghai Calling                    9.0  
5042                                 My Date with Drew                   84.0  

[5043 rows x 3 columns]

Kemudian kita cek dataset kita yang baru, terlihat seperti contoh di atas. Dataset jadi hanya 3 kolom yang tadinya berjumlah 28 kolom.


*   genres : Genre dari masing-masing film.
*   movie_title : Berisi judul-judul film.
*   num_user_for_reviews : Berisi review Users atas film tersebut.






In [34]:
print("Jumlah data rewview users adalah: ", len(movie_data.num_user_for_reviews.unique()))
print("Jumlah data nama movie adalah: ", len(movie_data.movie_title.unique()))
print("Jumlah data nama genre adalah: ", len(movie_data.genres.unique()))

Jumlah data rewview users adalah:  955
Jumlah data nama movie adalah:  4917
Jumlah data nama genre adalah:  914


Kita juga melihat panjang dari data dengan fungsi unique() agar menghindari data dengan hitungan duplikat.

# Data Preparation

Kita tidak melakukan tahap Data Preprocessing karena kita hanya memiliki satu file utuh, kita tidak melakukan penggabungan data pada kasus ini. Jadi kita langsung mulai ke tahap Data Preparation

In [35]:
movie_data.isnull().sum()

genres                   0
movie_title              0
num_user_for_reviews    21
dtype: int64

Sebelumnya kita melihat pada data utuh banyak sekali kita mendapati Missing Value, namun kita juga akan memeriksanya kembali dengan fungsi yang sama.

Dan setelah data di pisah, kemudian kita melihat Missing Value pada data, terlihat hanya atribut num_user_for_reviews yang memiliki Missing Value dengan total Missing 21 baris.

In [36]:
movie_data = movie_data.dropna()

Kita akan menghapus data Missing tersebut dengan fungsi dropna(), dengan fungsi tersebut kita bisa otomatis membuang baris pada kolom yang mengandung Missing Value.

In [37]:
movie_data.isnull().sum()

genres                  0
movie_title             0
num_user_for_reviews    0
dtype: int64

Setelah kita cek kembali data kita dengan isnull().sum(), data kita telah bersih dari Missing Value atau nilai kosong.

In [38]:
# Mengurutkan movie_data berdasarkan Review Users
movie_data = movie_data.sort_values('num_user_for_reviews', ascending=True)
movie_data

genres  \
2349          Biography|Drama|History   
4990                      Documentary   
4774          Comedy|Mystery|Thriller   
4771                      Crime|Drama   
4918               Comedy|Drama|Music   
...                               ...   
240   Action|Adventure|Fantasy|Sci-Fi   
654                     Action|Sci-Fi   
1937                      Crime|Drama   
66        Action|Crime|Drama|Thriller   
270    Action|Adventure|Drama|Fantasy   

                                            movie_title  num_user_for_reviews  
2349                                         Ramanujan                    1.0  
4990                                          Counting                    1.0  
4774                          The Sound and the Shadow                    1.0  
4771                                The King of Najayo                    1.0  
4918                               Theresa Is a Mother                    1.0  
...                                                 ...                   ...  
240          Star Wars: Episode I - The Phantom Menace                 3597.0  
654                                         The Matrix                 3646.0  
1937                          The Shawshank Redemption                 4144.0  
66                                     The Dark Knight                 4667.0  
270   The Lord of the Rings: The Fellowship of the R...                5060.0  

[5022 rows x 3 columns]

Setelah kita urutkan dengan fungsi sort_values(), bisa dilihat pada kolom num_user_for_reviews bahwa data di urutkan dari yang terkecil hingga yang terbesar. 
Dengan review yang paling sedikit sebanyak 1 sampai yang terbanyak yaitu 5060 review

In [39]:
len(movie_data.num_user_for_reviews.unique())

954

Kita coba cek data kita dengan memfilterya dengan unique(), jumlah nya adalah 954.

In [40]:
movie_data = movie_data.drop_duplicates('num_user_for_reviews')
movie_data

genres  \
2349          Biography|Drama|History   
206                           Fantasy   
4201                            Drama   
4711                 Mystery|Thriller   
4782                           Action   
...                               ...   
240   Action|Adventure|Fantasy|Sci-Fi   
654                     Action|Sci-Fi   
1937                      Crime|Drama   
66        Action|Crime|Drama|Thriller   
270    Action|Adventure|Drama|Fantasy   

                                            movie_title  num_user_for_reviews  
2349                                         Ramanujan                    1.0  
206       Harry Potter and the Deathly Hallows: Part I                    2.0  
4201                     Valley of the Heart's Delight                    3.0  
4711                                      Arnolds Park                    4.0  
4782                                      Diamond Ruff                    5.0  
...                                                 ...                   ...  
240          Star Wars: Episode I - The Phantom Menace                 3597.0  
654                                         The Matrix                 3646.0  
1937                          The Shawshank Redemption                 4144.0  
66                                     The Dark Knight                 4667.0  
270   The Lord of the Rings: The Fellowship of the R...                5060.0  

[954 rows x 3 columns]

Setelah itu kita akan menghapus juga data agar tidak memiliki duplikat data, kita menghapuskan data berdasarkan review users.

In [41]:
# konversi data series menjadi list

genres = movie_data['genres'].tolist()

movie = movie_data['movie_title'].tolist()

review_user = movie_data['num_user_for_reviews'].tolist()

Selanjutnya kita mengkonversikan data kita menjadi data list, dengan fungsi tolist() data kita akan di jadikan data list.

In [42]:
print(len(genres))
print(len(movie))
print(len(review_user))

954
954
954


Kita juga ingin cek jumlah total data kita dari masing-masing atribut, setiap atribute memiliki jumlah yang sama yaitu 954, Bagus!

In [43]:
movie

['Ramanujan\xa0',
 'Harry Potter and the Deathly Hallows: Part I\xa0',
 "Valley of the Heart's Delight\xa0",
 'Arnolds Park\xa0',
 'Diamond Ruff\xa0',
 'House at the End of the Drive\xa0',
 'Hidden Away\xa0',
 'Tiger Orange\xa0',
 'The Helix... Loaded\xa0',
 '3 Strikes\xa0',
 'Hansel & Gretel Get Baked\xa0',
 'Restless\xa0',
 'Green Street 3: Never Back Down\xa0',
 'Chicken Tikka Masala\xa0',
 'On the Outs\xa0',
 'The Blue Room\xa0',
 'Unnatural\xa0',
 'Queen Crab\xa0',
 'The Magic Flute\xa0',
 'Detention of the Dead\xa0',
 'I Married a Strange Person!\xa0',
 "A Turtle's Tale: Sammy's Adventures\xa0",
 'Woman Thou Art Loosed\xa0',
 'Ira & Abby\xa0',
 'Gomorrah\xa0            ',
 "She's Gotta Have It\xa0",
 'Out of the Blue\xa0',
 'Shanghai Surprise\xa0',
 'The Valley of Decision\xa0',
 'Chocolate: Deep Dark Secrets\xa0',
 'Khiladi 786\xa0',
 'Hollywood Shuffle\xa0',
 'The Sisterhood of the Traveling Pants 2\xa0',
 'The Christmas Candle\xa0',
 'Aberdeen\xa0',
 'Benji\xa0',
 'Beyond the 

Namun kita memiliki problem pada kasus kita saat membuat sistem rekomendasi, yaitu tiba-tiba data film kita setelah di jadikan list menjadi berubah. Bisa kita lihat di atas nama setiap film menjadi ada tambahan di belakangnya yaitu **\ax0**.

Ini akan akan sangat mengganggu ketika kita ingin mencari rekomendasi film yang serupa berdasarkan genre, karena kita harus memasukan **\ax0** setelah menulis judul film, kita tidak menginginkan itu.

Jadi kita putuhkan untuk menghapus tulisan tersebut dari data film kita.

In [44]:
movie = list(map(lambda st: str.replace(st, "\xa0", ""), movie))
movie

['Ramanujan',
 'Harry Potter and the Deathly Hallows: Part I',
 "Valley of the Heart's Delight",
 'Arnolds Park',
 'Diamond Ruff',
 'House at the End of the Drive',
 'Hidden Away',
 'Tiger Orange',
 'The Helix... Loaded',
 '3 Strikes',
 'Hansel & Gretel Get Baked',
 'Restless',
 'Green Street 3: Never Back Down',
 'Chicken Tikka Masala',
 'On the Outs',
 'The Blue Room',
 'Unnatural',
 'Queen Crab',
 'The Magic Flute',
 'Detention of the Dead',
 'I Married a Strange Person!',
 "A Turtle's Tale: Sammy's Adventures",
 'Woman Thou Art Loosed',
 'Ira & Abby',
 'Gomorrah            ',
 "She's Gotta Have It",
 'Out of the Blue',
 'Shanghai Surprise',
 'The Valley of Decision',
 'Chocolate: Deep Dark Secrets',
 'Khiladi 786',
 'Hollywood Shuffle',
 'The Sisterhood of the Traveling Pants 2',
 'The Christmas Candle',
 'Aberdeen',
 'Benji',
 'Beyond the Lights',
 'The Basket',
 'Light It Up',
 'Blackthorn',
 'Keeping Up with the Steins',
 'The 41-Year-Old Virgin Who Knocked Up Sarah Marshall and

Dengan lamda dan melakukan replace pada data film, kita sudah membuang tulisan yang mengganggu tersebut. Sekarang data kita sudah benar-benar bersih dari gangguan.

In [45]:
movie_new = pd.DataFrame({
    'review_users' : review_user,
    'movie' : movie,
    'genre' : genres
})
movie_new

review_users                                              movie  \
0             1.0                                          Ramanujan   
1             2.0       Harry Potter and the Deathly Hallows: Part I   
2             3.0                      Valley of the Heart's Delight   
3             4.0                                       Arnolds Park   
4             5.0                                       Diamond Ruff   
..            ...                                                ...   
949        3597.0          Star Wars: Episode I - The Phantom Menace   
950        3646.0                                         The Matrix   
951        4144.0                           The Shawshank Redemption   
952        4667.0                                    The Dark Knight   
953        5060.0  The Lord of the Rings: The Fellowship of the Ring   

                               genre  
0            Biography|Drama|History  
1                            Fantasy  
2                              Drama  
3                   Mystery|Thriller  
4                             Action  
..                               ...  
949  Action|Adventure|Fantasy|Sci-Fi  
950                    Action|Sci-Fi  
951                      Crime|Drama  
952      Action|Crime|Drama|Thriller  
953   Action|Adventure|Drama|Fantasy  

[954 rows x 3 columns]

Selanjutnya kita akan membuat Dictionary dari data yang sudah kita perbaiki sebelumnya untuk menentukan key:value pada setiap atribut.

# Model Development with Content Based Filtering

Sampai di mana kita pada tahapan Model Development menggunakan metode Based Filtering.

In [46]:
data = movie_new
data.sample(10)

review_users                        movie  \
293         294.0  Mad Max 2: The Road Warrior   
381         383.0                     Red Dawn   
746         904.0                       Frozen   
602         644.0             The Black Dahlia   
365         367.0                Lilo & Stitch   
375         377.0             Moonrise Kingdom   
363         365.0                        Glory   
433         436.0                The Equalizer   
713         836.0                      Boyhood   
508         527.0             Captain Phillips   

                                                 genre  
293                   Action|Adventure|Sci-Fi|Thriller  
381                                   Action|Drama|War  
746  Adventure|Animation|Comedy|Family|Fantasy|Musical  
602                       Crime|Drama|Mystery|Thriller  
365  Adventure|Animation|Comedy|Drama|Family|Fantas...  
375                     Adventure|Comedy|Drama|Romance  
363                                  Drama|History|War  
433                              Action|Crime|Thriller  
713                                              Drama  
508                           Biography|Drama|Thriller

Kita membuat variable baru bernama **data** lalu melihat sample acak sebanyak 10 baris.

Kita akan membangun model sistem rekomendasi sederhana film berdasarkan genrenya, teknik yang kita gunakan adalah dengan TF-IDF Vectorizer.

In [47]:
from sklearn.feature_extraction.text import TfidfVectorizer
 
# Inisialisasi TfidfVectorizer
tf = TfidfVectorizer()
 
# Melakukan perhitungan idf pada data genre
tf.fit(data['genre']) 
 
tf.get_feature_names() 

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['action',
 'adventure',
 'animation',
 'biography',
 'comedy',
 'crime',
 'documentary',
 'drama',
 'family',
 'fantasy',
 'fi',
 'history',
 'horror',
 'music',
 'musical',
 'mystery',
 'romance',
 'sci',
 'sport',
 'thriller',
 'war',
 'western']

Kemudian kita fit_tranform data genre dalam bentuk matrik.

In [48]:
# Melakukan fit lalu ditransformasikan ke bentuk matrix
tfidf_matrix = tf.fit_transform(data['genre']) 
 
# Melihat ukuran matrix tfidf
tfidf_matrix.shape 

(954, 22)

Kita bisa melihatnya dengan fungsi todense()

In [49]:
# Mengubah vektor tf-idf dalam bentuk matriks dengan fungsi todense()
tfidf_matrix.todense()

matrix([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        ...,
        [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
         0.        ],
        [0.47834689, 0.        , 0.        , ..., 0.47399518, 0.        ,
         0.        ],
        [0.4734089 , 0.51386057, 0.        , ..., 0.        , 0.        ,
         0.        ]])

Selanjutnya, kita akan lihat matriks tf-idf untuk beberapa nama film dan kategori genrenya

In [50]:
# Membuat dataframe untuk melihat tf-idf matrix
 
pd.DataFrame(
    tfidf_matrix.todense(), 
    columns=tf.get_feature_names(),
    index=data.movie
).sample(22, axis=1).sample(10, axis=0)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


mystery  western  sport  family  \
movie                                                                       
Diary of a Mad Black Woman               0.000000      0.0    0.0  0.0000   
Pale Rider                               0.000000      1.0    0.0  0.0000   
What Dreams May Come                     0.000000      0.0    0.0  0.0000   
Miracle at St. Anna                      0.000000      0.0    0.0  0.0000   
King Kong                                0.000000      0.0    0.0  0.0000   
Ghost Rider                              0.000000      0.0    0.0  0.0000   
Killer Joe                               0.000000      0.0    0.0  0.0000   
The Hurt Locker                          0.000000      0.0    0.0  0.0000   
Red Planet                               0.000000      0.0    0.0  0.0000   
Harry Potter and the Chamber of Secrets  0.529872      0.0    0.0  0.5841   

                                          fantasy     crime  music  thriller  \
movie                                                                          
Diary of a Mad Black Woman               0.000000  0.000000    0.0  0.000000   
Pale Rider                               0.000000  0.000000    0.0  0.000000   
What Dreams May Come                     0.642704  0.000000    0.0  0.000000   
Miracle at St. Anna                      0.000000  0.469945    0.0  0.353827   
King Kong                                0.000000  0.000000    0.0  0.000000   
Ghost Rider                              0.671490  0.000000    0.0  0.521576   
Killer Joe                               0.000000  0.587026    0.0  0.441978   
The Hurt Locker                          0.000000  0.000000    0.0  0.332838   
Red Planet                               0.000000  0.000000    0.0  0.439843   
Harry Potter and the Chamber of Secrets  0.468295  0.000000    0.0  0.000000   

                                         documentary    action  ...  \
movie                                                           ...   
Diary of a Mad Black Woman                       0.0  0.000000  ...   
Pale Rider                                       0.0  0.000000  ...   
What Dreams May Come                             0.0  0.000000  ...   
Miracle at St. Anna                              0.0  0.357075  ...   
King Kong                                        0.0  0.471896  ...   
Ghost Rider                                      0.0  0.526364  ...   
Killer Joe                                       0.0  0.000000  ...   
The Hurt Locker                                  0.0  0.000000  ...   
Red Planet                                       0.0  0.443881  ...   
Harry Potter and the Chamber of Secrets          0.0  0.000000  ...   

                                         biography  horror  adventure  \
movie                                                                   
Diary of a Mad Black Woman                     0.0     0.0   0.000000   
Pale Rider                                     0.0     0.0   0.000000   
What Dreams May Come                           0.0     0.0   0.000000   
Miracle at St. Anna                            0.0     0.0   0.000000   
King Kong                                      0.0     0.0   0.512218   
Ghost Rider                                    0.0     0.0   0.000000   
Killer Joe                                     0.0     0.0   0.000000   
The Hurt Locker                                0.0     0.0   0.000000   
Red Planet                                     0.0     0.0   0.000000   
Harry Potter and the Chamber of Secrets        0.0     0.0   0.398451   

                                          romance    comedy        fi  \
movie                                                                   
Diary of a Mad Black Woman               0.688798  0.580974  0.000000   
Pale Rider                               0.000000  0.000000  0.000000   
What Dreams May Come                     0.648340  0.000000  0.000000   
Miracle at St. Anna                      0.000000  0.000000  

Kita bisa lihat hasil di atas, sebagai contoh pada film Grown Ups memiliki sekitar 1.0 bergenre Comedy.

Dari sini kita sudah berhasil merepresantasikan fitur penting pada kategori genre, dan menunjukan korelasi film terhadap genrenya. 

Kemudian kita akan menghitung derajat kesamaan antar film dengan teknik Cosine Similarity, kita akan menggunakan library scikit-learn untuk menggunakan teknik Cosine Similarity

In [51]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
genre_sim = cosine_similarity(tfidf_matrix) 
genre_sim

array([[1.        , 0.        , 0.28987898, ..., 0.15195737, 0.11233677,
        0.11117712],
       [0.        , 1.        , 0.        , ..., 0.        , 0.        ,
        0.60393419],
       [0.28987898, 0.        , 1.        , ..., 0.52420967, 0.38752989,
        0.38352942],
       ...,
       [0.15195737, 0.        , 0.52420967, ..., 1.        , 0.73926506,
        0.20104983],
       [0.11233677, 0.        , 0.38752989, ..., 0.73926506, 1.        ,
        0.37508279],
       [0.11117712, 0.60393419, 0.38352942, ..., 0.20104983, 0.37508279,
        1.        ]])

Kemudian kita akan melihat hasil dari kesamaan antar film yang telah kita lakukan pada teknik Cosine Similarity.

In [52]:
# Membuat dataframe dari variabel cosine_sim dengan baris dan kolom berupa nama film
genre_sim_df = pd.DataFrame(genre_sim, index=data['movie'], columns=data['movie'])
print('Shape:', genre_sim_df.shape)
 
# Melihat similarity matrix pada setiap film
genre_sim_df.sample(5, axis=1).sample(10, axis=0)

Shape: (954, 954)


movie                    Underworld: Evolution  Cloud Atlas   Ant-Man  \
movie                                                                   
Cinderella                            0.227879     0.133232  0.000000   
Click                                 0.256249     0.149819  0.205064   
Michael Clayton                       0.149767     0.145133  0.000000   
Charlie St. Cloud                     0.292061     0.170757  0.000000   
Lars and the Real Girl                0.000000     0.181413  0.248307   
The Blair Witch Project               0.000000     0.000000  0.000000   
American Psycho                       0.000000     0.219312  0.000000   
The Game Plan                         0.000000     0.000000  0.159180   
Hamlet 2                              0.000000     0.000000  0.188620   
Avatar                                0.935634     0.608199  0.790271   

movie                    Snakes on a Plane   Sicario  
movie                                                 
Cinderella                        0.109528  0.101555  
Click                             0.123163  0.114198  
Michael Clayton                   0.612674  0.919269  
Charlie St. Cloud                 0.140376  0.130158  
Lars and the Real Girl            0.149136  0.138280  
The Blair Witch Project           0.000000  0.000000  
American Psycho                   0.656098  0.608339  
The Game Plan                     0.000000  0.000000  
Hamlet 2                          0.000000  0.000000  
Avatar                            0.352056  0.149862

Yang perlu di perhatikan kenapa kesamaan nya tidak mencapai angka 1.0 adalah karena kita memiliki film dengan genre lebih dari satu. Namun kalau kita mencoba mencari secara lengkap ada film dengan kesamaan 1.0. 

Jadi kita coba merekomendasikan sebuah film kepada Users berdasarkan genrenya, dan sebuah genre pun bermacam macam.

## Melihat Hasil Rekomendasi

In [53]:
def film_recommendations(nama_movie, similarity_data=genre_sim_df, items=data[['movie', 'genre']], k=5):
    
    index = similarity_data.loc[:,nama_movie].to_numpy().argpartition(
        range(-1, -k, -1))
    
    closest = similarity_data.columns[index[-1:-(k+2):-1]]
    closest = closest.drop(nama_movie, errors='ignore')
 
    return pd.DataFrame(closest).merge(items).head(k)

Kita akan coba masukan film Avatar, jika kita lihat Avatar memiliki 4 genre yaitu Action, Adventure, Fantasy, dan Sci-Fi. Kita akan mencoba merekomdasikan sebuah film dari genre tersebut.

In [54]:
data[data.movie.eq('Avatar')]

review_users   movie                            genre
944        3054.0  Avatar  Action|Adventure|Fantasy|Sci-Fi

In [55]:
film_recommendations('Avatar')

movie  \
0    Star Wars: Episode VI - Return of the Jedi   
1                      The Amazing Spider-Man 2   
2                                  Man of Steel   
3  Star Wars: Episode III - Revenge of the Sith   
4                           Highlander: Endgame   

                             genre  
0  Action|Adventure|Fantasy|Sci-Fi  
1  Action|Adventure|Fantasy|Sci-Fi  
2  Action|Adventure|Fantasy|Sci-Fi  
3  Action|Adventure|Fantasy|Sci-Fi  
4  Action|Adventure|Fantasy|Sci-Fi

Menarik! Ternyata ada film dengan 4 genre yang sama, tentu dengan nama film yang berbeda. Jadi kita akan merekomendasikan film tersebut kepada Users, barangkali tertarik.